textdb/textdb-textql/src/main/javacc/TextQLParser.jj

```diff
@@ -0,0 +1,697 @@
+        }
+        public KeywordExtractPredicate(List<String> matchingFields, String keywords, String matchingType) {
+            this.matchingFields = matchingFields;
+            this.keywords = keywords;
+            this.matchingType = matchingType;
+        }
+        @Override
+        public boolean equals(Object other) {
+            if (other == null) { return false; }
+            if (other.getClass() != getClass()) { return false; }
+            KeywordExtractPredicate keywordExtractPredicate = (KeywordExtractPredicate) other;
+            return new EqualsBuilder()
+                    .appendSuper(super.equals(keywordExtractPredicate))
+                    .append(matchingFields, keywordExtractPredicate.matchingFields)
+                    .append(keywords, keywordExtractPredicate.keywords)
+                    .append(matchingType, keywordExtractPredicate.matchingType)
+                    .isEquals();
+        }
+    }
+
+    /**
+     * Generates a unique deterministic sequence of ids for new statements
+     * The id follows the format "_sidX", where X is an incremental hexadecimal
+     *  integer without padding starting from 0.
+     * e.g. "_lid0", "_lid1", ..., "_lid9", "_lida", ..., "_lidf", "_lid10", ...
+     * @return The generated id
+     */
+    private int currentStatementIdSequence = 0;
+    private String getNewStatementId(){
+        return "_sid" + Integer.toString(currentStatementIdSequence++, 16);
+    }
+}
+
+PARSER_END(TextQLParser)
+
+/*
+ *    Skipped Tokens
+ *  The following are tokens that should be ignored/skipped.
+ */
+SKIP : { " " | "\n" | "\r" | "\t" | "\f" }
+/**
+ * Basic Tokens - Reserved Words
+ */
+TOKEN : { < CREATE : "CREATE" > }
+TOKEN : { < DIRECTORY : "DIRECTORY" > }
+TOKEN : { < VIEW : "VIEW" > }
+TOKEN : { < SELECT : "SELECT" > }
+TOKEN : { < EXTRACT : "EXTRACT" > }
+TOKEN : { < FROM : "FROM" > }
+TOKEN : { < AS : "AS" > }
+TOKEN : { < LIMIT : "LIMIT" > }
+TOKEN : { < OFFSET : "OFFSET" > }
+TOKEN : { < JOIN : "JOIN" > }
+TOKEN : { < WHERE : "WHERE" > }
+TOKEN : { < OUTPUT : "OUTPUT" > }
+TOKEN : { < INSERT : "INSERT" > }
+TOKEN : { < INTO : "INTO" > }
+TOKEN : { < VALUES : "VALUES" > }
+TOKEN : { < KEYWORDMATCH : "KEYWORDMATCH" > }
+/**
+ *    Basic Tokens - Comparison operators
+ */
+TOKEN : { < ASSIGN: "="> }
+TOKEN : { < LT: "<" > }
+TOKEN : { < GT: ">" > }
+TOKEN : { < EQ: "==" > }
+TOKEN : { < LE: "<=" > }
+TOKEN : { < GE: ">=" > }
+TOKEN : { < NE: "!=" > }
+/**
+ * Basic Tokens - Other 1-Character tokens
+ */
+TOKEN : { < LPAREN: "(" > }
+TOKEN : { < RPAREN: ")" > }
+TOKEN : { < LBRACE: "{" > }
+TOKEN : { < RBRACE: "}" > }
+TOKEN : { < LBRACKET: "[" > }
+TOKEN : { < RBRACKET: "]" > }
+TOKEN : { < SEMICOLON: ";" > }
+TOKEN : { < COMMA: "," > }
+TOKEN : { < DOT: "." > }
+TOKEN : { < STAR : "*" > }
+
+
+
+/*
+ *    Literal Tokens
+ */
+/**
+ *    NUMBER_LITERAL - Matches a number in decimal format
+ *  The token matched is composed by an optional ('-') character followed by
```

Comment:
"The token matched is composed by an optional" -> "The matched token consists of an optional"